core/src/main/scala/cats/collections/Predicate.scala

```diff
@@ -46,28 +81,77 @@ abstract class Predicate[-A] extends scala.Function1[A, Boolean] { self =>
   /**
    * Return the opposite predicate
    */
-  def negate: Predicate[A] = Predicate(a => !apply(a))
+  def negate: Predicate[A]
 
   /**
    * Return the opposite predicate
    */
   def unary_!(): Predicate[A] = negate
+
+  /**
+   * Compose the predicate with a function.
+   *
+   * A value is a member of the resulting predicate iff its image through f is a
+   * member of this predicate.
+   */
+  def contramap[B](f: B => A): Predicate[B]
+
+  /**
+   * Alias for contramap.
+   */
+  final override def compose[B](f: B => A): Predicate[B] = contramap(f)
 }
 
 object Predicate extends PredicateInstances {
-  def apply[A](f: A => Boolean): Predicate[A] = new Predicate[A] {
-    def apply(a: A) = f(a)
+  private val True = Kleisli.liftF(Eval.True)
+  private val False = Kleisli.liftF(Eval.False)
+
+  private[collections] case object Empty extends Predicate[Any] {
+    override def run: Kleisli[Eval, Any, Boolean] = Predicate.False
+    override def negate: Predicate[Any] = Everything
+    override def contramap[B](f: B => Any): Predicate[B] = this
   }
 
-  def empty: Predicate[Any] = apply(_ => false)
+  private[collections] case object Everything extends Predicate[Any] {
+    override def run: Kleisli[Eval, Any, Boolean] = Predicate.True
+    override def negate: Predicate[Any] = Empty
+    override def contramap[B](f: B => Any): Predicate[B] = this
+  }
+
+  private[collections] final case class Lift[A](run: Kleisli[Eval, A, Boolean]) extends Predicate[A] {
+    override def negate: Predicate[A] = Negate(this)
+    override def contramap[B](f: B => A): Predicate[B] = Lift(run.compose(b => Eval.now(f(b))))
+  }
+
+  private[collections] final case class Negate[A](negate: Predicate[A]) extends Predicate[A] {
+    override def run: Kleisli[Eval, A, Boolean] = negate.run.map(!_)
+    override def contramap[B](f: B => A): Predicate[B] = Negate(negate contramap f)
+  }
+
+  /**
+   * build a set from a membership function.
+  */
+  def apply[A](p: A => Boolean): Predicate[A] = Lift {
+    Kleisli(a => Eval.now(p(a)))
```

Comment:
I think `a => if (p(a)) Eval.True else Eval.False` may be marginally more efficient.